zerver/actions/message_send.py

```diff
@@ -104,6 +105,17 @@
 from zerver.tornado.django_api import send_event_on_commit
 
 
+def message_is_idempotent(user: UserProfile, local_id: str, queue_id: str) -> bool:
+    sent_message_local_id = f"{user.realm_id}:{user.id}:<{queue_id}>:{local_id}"
```

Comment:
`check_send_message` optionally gets a `realm` object as input. Could this object's `realm.id` be different from `sender.realm_id`?

If so, should we prefer the `realm` sent as input for this key?

Reply:
Yes, the sender can be a system bot not in the target realm. While system bots are unlikely to use this API ... one should just pass `realm` as its own parameter like we do in the caller.

Reply:
`realm.id` and `sender.realm_id` are the same (not always though), so I will just use the `realm.id` in the key if it's not `None` else I will use `sender.realm_id`